text/0071-trailing-commas.md

```diff
@@ -0,0 +1,187 @@
+# Trailing Commas
+
+## Related issues and PRs
+
+- Reference Issues: https://github.com/cedar-policy/cedar/issues/1024
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-06-26 
+
+## Summary
+
+The Cedar grammar (both policies and schemas) should accept trailing
+commas whenever possible.
+
+## Basic example
+
+Currently, this is a parse error:
+```
+permit(principal, action, resource) when {
+    { foo : 3, bar : 2, } == context  
+};
+```
+This RFC would make it not a parse error. 
+It would also change this is several places beyond records.
+
+## Motivation
+
+In general, allowing trailing commas makes editing easier.
+Re-arranging, deleting, and copy-pasting code into/from a comma
+separated grammar element will no longer require adjusting commas.
+
+Example:
+```
+{
+    "managers": ["Dave", "Shelly", "Chris"],
+    "projects" : {
+        "Alpha" : {
+            "zones" : ["A", "B"]
+        },
+        "Beta" : {
+            "zones" : ["D", "F"]
+        }
+    }
+}
+```
+The `Beta` project cannot simply be deleted from this record, and
+likewise, new fields cannot be simply pasted into any level of the
+objects, without adjusting commas.
+
+Many languages allow for trailing commas, including:
+* Rust
+* Python
+* Java,
+* Go
+* JavaScript
+* OCaml
+
+(Frustratingly: not JSON!)
+
+In addition, it will make our schema grammar and policy grammar more
+consistent, as the schema grammar allows trailing commas in many places.
+
+## Detailed design
+
+### Cedar Policy Grammar
+
+We want to allow trailing commas in the following expressions:
+1. scopes: `permit(principal, action, resource, )` 
+2. sets: `[1,2,3,]`
+3. records: `{ foo : 1, bar : 3, }`
+4. all method/function calls: `.contains(3,)`, `ip("10.10.10.10",)`
+
+In the grammar:
+
+The following grammar rules change:
+```
+Scope ::= Principal ',' Action ',' Resource
+```
+becomes:
+```
+Scope ::= Principal ',' Action ',' Resource ','?
+```
+
+```
+EntList ::= Entity {',' Entity}
+```
+becomes:
+```
+EntList ::= Entity {',' | ',' Entity} ','?
```

Comment:
```suggestion
EntList ::= Entity {',' | ',' Entity}
```

I think you don't want this extra `,?` because it would allow extra commas (e.g., `User::"alice", User::"bob",,`). Same comment for a few other places in the rfc.

Reply:
lol, i think grammar as written would allow as many commas as you like between entities `User::"alice",,,,,,,,,,,,,User::"bob"`